### PR TITLE
MGMT-9672: align manifest authz

### PR DESCRIPTION
**Pull request by slaviered**
The manifest handler implements its own access checks within the application layer. Those verifications are redundant and with the new tenancy policies also incorrect.

Since manifests are sub-objects of clusters, their access should is governed by the cluster access rules in the auth layer which should be the sole responsible for authorization checks.

Tests were changed accordingly and the case of missing cluster (which is handled by the auth middleware) was moved to the subsystem tests.

## List all the issues related to this PR

- [x] New Feature 
- [ ] Enhancement 
- [x] Bug fix
- [ ] Tests
- [ ] Documentation
- [ ] CI/CD 

## What environments does this code impact?

- [ ] Automation (CI, tools, etc)
- [x] Cloud
- [ ] Operator Managed Deployments
- [ ] None

## How was this code tested?

- [ ] assisted-test-infra environment
- [ ] dev-scripts environment
- [x] Reviewer's test appreciated
- [ ] Waiting for CI to do a full test run
- [ ] Manual (Elaborate on how it was tested)
- [ ] No tests needed

## Assignees

/cc @
/cc @

## Checklist

- [ ] Title and description added to both, commit and PR.
- [ ] Relevant issues have been associated (see [CONTRIBUTING] guide)
- [ ] Reviewers have been listed
- [ ] This change does not require a documentation update (docstring, `docs`, README, etc)
- [ ] Does this change include unit-tests (note that code changes require unit-tests)

## Reviewers Checklist

- Are the title and description (in both PR and commit) meaningful and clear?
- Is there a bug required (and linked) for this change?
- Should this PR be backported?

[Kubernetes community documentation]: https://github.com/kubernetes/community/blob/master/contributors/guide/pull-requests.md#commit-message-guidelines
[CONTRIBUTING]: https://github.com/openshift/assisted-service/blob/master/CONTRIBUTING.md

### CODE REVIEW

**Comment by slaviered** [comment]
cc: @danielerez 
cc: @carbonin

File: subsystem/manifests_test.go [hunk 1]
```diff
@@ -138,6 +139,44 @@ spec:
 		})
 	})
 
+	It("cluster doesn't exist", func() {
+		non_exiting_id := strfmt.UUID(uuid.New().String())
+		By("create manifest returns not found")
```
Comment (danielerez) [inline subsystem/manifests_test.go:144]: for clarity, maybe better to put each of these in a separate By func. E.g as done in https://github.com/openshift/assisted-service/pull/3510/files#diff-6173872f98a7065d377afcb41777af74922ed6d692ac80254da515c6d3477d88L153

Comment (slaviered) [reply subsystem/manifests_test.go:144]: It has the same effect

File: subsystem/manifests_test.go [hunk 2]
```diff
@@ -138,6 +139,44 @@ spec:
...
+		_, err := userBMClient.Manifests.V2CreateClusterManifest(ctx, &manifests.V2CreateClusterManifestParams{
+			ClusterID: non_exiting_id,
+			CreateManifestParams: &models.CreateManifestParams{
+				Content:  &base64Content,
+				FileName: &manifestFile.FileName,
+				Folder:   &manifestFile.Folder,
+			},
+		})
+		Expect(err).To(BeAssignableToTypeOf(installer.NewGetInfraEnvNotFound()))
```
Comment (carbonin) [inline subsystem/manifests_test.go:153]: I don't understand this. Why would we return an infraEnv error for a cluster-related API?

File: subsystem/manifests_test.go (hunk 1 above)
Comment (carbonin) [reply subsystem/manifests_test.go:144]: Yes, but the test output is more clear when they are separate. Also then you'll know if one or multiple are failing.
Currently if any expectation here fails then the rest of the tests don't get run.

**Review by carbonin — CHANGES_REQUESTED** [review]
Please separate out the test cases into their own `It` blocks for separate endpoints.

File: subsystem/manifests_test.go (hunk 2 above)
Comment (danielerez) [reply subsystem/manifests_test.go:153]: oh right, it should probably be NewV2GetClusterNotFound

File: internal/manifests/manifests_test.go [hunk 3]
```diff
@@ -198,20 +198,6 @@ var _ = Describe("ClusterManifestTests", func() {
 			Expect(responsePayload.Payload.Folder).To(Equal(defaultFolder))
 		})
 
-		It("cluster doesn't exist", func() {
```
Comment (carbonin) [inline internal/manifests/manifests_test.go:201]: Why remove these tests? Shouldn't we still error in these cases?

Comment (slaviered) [reply internal/manifests/manifests_test.go:201]: As I wrote in the PR description, the logic is not in the app layer anymore but in the middleware so those tests were moved to the subsystem test to make sure the system behaves as we expect when we depends on the authz handler

File: subsystem/manifests_test.go (hunk 2 above)
Comment (slaviered) [reply subsystem/manifests_test.go:153]: Yes. It was an copy paste mistake. Fixing

File: subsystem/manifests_test.go (hunk 1 above)
Comment (slaviered) [reply subsystem/manifests_test.go:144]: > Please separate out the test cases into their own `It` blocks for separate endpoints.
Each steps in the flow depends on the other (we first create, then list, and only at the end delete). This is why it has to be done within a single It (the order in a context is not guaranteed)

Comment (carbonin) [reply subsystem/manifests_test.go:144]: If the calls succeeded then sure, but there is no dependency when every call errors.
And either way I would still suggest they be separate.

File: internal/manifests/manifests_test.go (hunk 3 above)
Comment (slaviered) [reply internal/manifests/manifests_test.go:201]: we have but not here but in the subsystem. These function will not be called if the cluster does not exist

Comment (carbonin) [reply internal/manifests/manifests_test.go:201]: > These function will not be called if the cluster does not exist

What about cases where we're running without auth or in kube-api?
The authz middleware doesn't check ownership in those scenarios.
We should have a sane response from these functions even if we're not using rhsso auth.

Comment (carbonin) [reply internal/manifests/manifests_test.go:201]: From @avishayt in a slack conversation:

> If the question is "do we need to make sure that the cluster exists?", I would say "nice to have but no".  In SaaS, if someone deleted the cluster then the authz won't allow them to view the manifests.  In k8s API, manifests are read from ConfigMaps and not from the backend anyway (the controller lists manifests just to check which ones it already created in the backend).

I guess this means that whether the cluster exists in the database is actually irrelevant (before and after this change). I suppose we could have a test for when the manifest doesn't exist in s3, but that doesn't seem super important.

File: subsystem/manifests_test.go (hunk 1 above)
Comment (carbonin) [reply subsystem/manifests_test.go:144]: FWIW I would still like these to be split.
There's no reason for them to be in one test and it should be pretty low effort to separate them

Comment (carbonin) [reply subsystem/manifests_test.go:144]: I don't understand why this keeps getting resolved.
Am I missing something with how this test works?

It really feels like there's no reason that these should be in the same test case.

Comment (slaviered) [reply subsystem/manifests_test.go]: test case are now separated to individual tests

**Comment by danielerez** [comment]
> link

/hold

**Pull request merged**